stablehlo/reference/Ops.cpp

```diff
@@ -265,10 +246,8 @@ Tensor evalPadOp(const Tensor &operand, const Tensor &paddingValue,
     result.set(*resultIt, paddingValue.get({}));
   for (auto operandIt = operand.index_begin(); operandIt != operand.index_end();
        ++operandIt) {
-    SmallVector<int64_t> resultIdx(result.getType().getRank());
-    for (auto i = 0; i < operand.getType().getRank(); ++i)
-      resultIdx[i] =
-          edgePaddingLow[i] + (*operandIt)[i] * (interiorPadding[i] + 1);
+    Index resultIdx =
+        edgePaddingLow + *operandIt * interiorPadding + *operandIt;
```

Comment:
If use the overload`Index::operator+(int64_t n)`, we can achieve the following.

```
 Index resultIdx =
        edgePaddingLow + (*operandIt) * (Index(interiorPadding) + 1);
```
I prefer the implemented one with distributed `*operandIt` over explicit `Index` casts. Please let me know what you think?

Reply:
The current code looks good to me. 

For what it's worth, I think it would be possible to enable `... * (interiorPadding + 1)` by replacing uses of `ArrayRef<int64_t>` in Ops.cpp with domain-specific types: 
  * BroadcastInDimOp::broadcastDimensions, ReverseOp::dimensions and TransposeOp::permutation would become `Axes`.
  * Various startIndices would become `Indices`.
  * Other ArrayRef<Int64_t> parameters would become `Sizes`.
  * We'd also introduce `Tensor::getShape` which would return `Sizes`.
  * We could also have a helper function that turns `ArrayRef<Tensor> startIndices` into `Indices`. 

Then we could define the appropriate operators and helpers. E.g. we shouldn't do anything with Axes, but we can add, multiply etc Sizes and Indices and mix them with each other. This way we'd be able to replace code like this:

```C++
Index adjustedStartIndices(startIndices.size());
for (size_t i = 0; i < startIndices.size(); ++i)
  adjustedStartIndices[i] = std::min(
      std::max(startIndices[i].get({}).getIntegerValue().getSExtValue(), 0l),
      operand.getType().getShape()[i] - sliceSizes[i]);
```

With code like this:

```C++
auto adjustedStartIndices = clamp(0, evalIndices(startIndices), operand.getShape() - sliceSizes);
```

Reply:
I couldn't resist and prototyped how this could look like: https://github.com/sdasgup3/stablehlo/pull/4.

Reply:
The prototype is finished. Please review the PR to your branch, and then we could potentially merge it to make it part of this PR?

Reply:
Thanks for the suggestion. Looks great to me. Merged and resolved the TODO to remove `Tensor(Type type)` constructor. 

Regarding the prototype:

>> E.g. we shouldn't do anything with Axes, but we can add, multiply etc Sizes and Indices and mix them with each other. 

Currently we are using `Sizes` in places like
```
... evalBroadcastInDimOp(.. {
   ...
Sizes operandIdx(operandShape.size());
  ...
```

Shouldn't we be using an `Index` type in the following places to improve readability. 
1. in places like the above.
2. `evalIndices` to return `Index`.
3. `IndexSpaceIterator::operator*` and `IndexSpaceIterator::operator->` to return `Index`.
4. The private member of `IndexSpaceIterator` to be `std::optional<Index> index_`.

 `Index` could be an alias of `Sizes` (i.e. `using Index = Sizes;`).

What do you think?
@burmako

Reply:
Yeah, I also don't love the current situation. This is why I initially had `Index` and `Sizes` as two different types (you can see in commit history), but that was a lot of boilerplate.

Eventually, I kind of convinced myself to follow LLVM's (and C++'s) convention to use `size_t` to represent both sizes and indices, which generalizes to `Sizes` and sizes/indices in StableHLO, so I merged `Index` and `Sizes` into just `Sizes`.

Let's give `use Index = Sizes;` a try and see how well this works out. I'd also add "6) Signatures of evalFooOp functions where appropriate" as another place to use `Index`.

Reply:
PTAL.
Added one more
7. Called `evalIndices` to prepare an `Index` to be passed into the following call to `evalDynamicSliceOp` or `evalDynamicUpdateSliceOp`. This allows  those `evalFoOp` to have `Index startIndices` as formal arguments which is more readable.